R/BackendHdf5-class.R

```diff
@@ -0,0 +1,225 @@
+        ))
+    }
+    NULL
+}
+
+.valid.BackendHdf5.h5files <- function(src, h5files) {
+    if (length(src) != length(h5files))
+        return("Different number of hdf5 and original source files.")
+    if (anyDuplicated(h5files))
+        return("Hdf5 files are not unique.")
+    if (!all(file.exists(h5files))) {
+        return(paste0(
+            "File(s) ",
+            paste0(h5files[!file.exists(h5files)], collapse=", "),
+            " not found."
+        ))
+    }
+    NULL
+}
+
+setValidity("BackendHdf5", function(object) {
+    msg <- c(
+        .valid.BackendHdf5.h5files(object@files, object@h5files),
+        .valid.BackendHdf5.checksums(object@checksums, object@h5files)
+    )
+    if (is.null(msg)) { TRUE } else { msg }
+})
+
+#' Initialize the Hdf5 backend, i.e. create the hdf5 files and add their names
+#' to the @hdf5file slot.
+#'
+#' @param object `BackendHdf5`
+#'
+#' @param files `character` with the original file names.
+#'
+#' @param spectraData `DataFrame` with the spectrum metadata.
+#'
+#' @param path `character(1)` defining the path where the hdf5 files should be
+#'     stored.
+#'
+#' @return `BackendHdf5` object.
+#'
+#' @author Johannes Rainer, Sebastian Gibb
+#'
+#' @md
+#'
+#' @noRd
+setMethod(
+    "backendInitialize",
+    "BackendHdf5",
+    function(object, files, spectraData, path=".", ...) {
+    path <- normalizePath(path, mustWork=FALSE)
+    dir.create(path, showWarnings=FALSE, recursive=TRUE)
+
+    files <- normalizePath(files)
+    object@h5files <- file.path(path, paste0(.vdigest(files), ".h5"))
+
+    if (any(file.exists(object@h5files))) {
+        stop(
+            "File(s) ",
+            paste0(object@h5files[file.exists(object@h5files)], collapse=", "),
+            " already exist(s). Please choose a different 'path'."
+        )
+    }
+
+    for (i in seq(along=object@h5files)) {
+        h5 <- H5Fcreate(object@h5files[i])
+        h5createGroup(h5, "spectra")
+        H5Fclose(h5)
+    }
+
+    object@checksums <- .vdigest(object@h5files, file=TRUE)
+
+    callNextMethod()
+})
+
+#' Import the data from the raw MS files and store them to the hdf5 files.
+#'
+#' @inheritParams backendInitialize
+#'
+#' @return `BackendHdf5`
+#'
+#' @author Johannes Rainer, Sebastian Gibb
+#'
+#' @md
+#'
+#' @noRd
+setMethod(
+    "backendImportData",
+    "BackendHdf5",
+    function(object, spectraData, ..., BPPARAM=bpparam()) {
+
+    bpmapply(object@files, object@h5files, FUN=.mzr2hdf5, BPPARAM=BPPARAM)
+    object@checksums <- .vdigest(object@h5files, file=TRUE)
+    validObject(object)
+    object
+})
+
+setMethod(
+    "backendReadSpectra",
+    "BackendHdf5",
+    function(object, spectraData, ...) {
+
+    uFileIdx <- unique(spectraData$fileIdx)
+    h5files <- object@h5files[uFileIdx]
+
+    changed <- .valid.BackendHdf5.checksums(object@checksums[uFileIdx], h5files)
+    if (!is.null(changed))
+        stop(changed)
+
+    spectra <- vector(mode="list", length=nrow(spectraData))
```

Comment:
Although I usually don't like having too many `if` conditions, I would make here a switch on whether data from a single or multiple files should be read:
```
if (length(h5files) == 1) {
  <code to read spectra from one file>
} else {
  <lapply/mapply over files>
}
```
The reason is that we want to make reading data as fast as possible and splitting, creating lists, assigning etc can slow stuff down. I tested that for the `BackendMzR` backend where I do also switch on the number of files `backendReadSpectra` should read data. Also, reading from multiple files should be the exception.